### PR TITLE
[NTOSKRNL][KE][INBV] Modern BSOD

### PR DESCRIPTION
**Pull request by feel-the-dz3n**
## Purpose

Seems like BSOD is the main sight. Make BSOD look like Windows 10.
### The goal is to provide user-friendly interface for users and more useful information to developers.
bsod 0

JIRA issue: [CORE-15474](https://jira.reactos.org/browse/CORE-15474)

## Proposed changes

- Added  `:(` image, palette: image
- Removed a lot of texts like `PSS_MESSAGE_INTRO`, `BUGCHECK_TECH_INFO`
- `BUGCHECK_MESSAGE_INTRO` is like in Windows 10 now: `Your PC ran into a problem and needs to restart.`
- `BUGCODE_ID_DRIVER` -> `What failed: `
- Added backtrace information: ThFabba@6a9f172
- Added ErrorScreenStyle to change style of BSOD (0 - modern + QR, 1 - modern, 2 - old + additional info)
- `AdditionalInformation` string will contain text for QR-code.
- You can see "Stop code:" on the screenshots, it's MessageId

## TODO

- [ ] Implement settings for `ErrorScreenStyle`.
- [ ] ~Make QR-codes (is it even possible? :thinking:)~ Fix QR-codes, they are not working anymore.
- [ ] Rebuild bugcodes.mc to show less text and more useful information. I need your help here, because I have bad English knowledge and I can't rebuild all strings by myself.
- [x] Make everything better!
- [ ] Make background color changeable (by creating gradient in palette?)
- [ ] Remove my comments from code when everything is done (i can't remember functions names and arguments :disappointed:)
 -  I've not seen all types of BSOD, so it may fail with some large texts. Please, send some screenshots if you found something.

# Let's make it together

I need your help to make it better. Please, send your edits/suggestions. For example, new strings for `bugcodes.mc`. **You can help with registry settings and QR code implementation.**

Here is `bootcd.iso` for tests:
- [**bootcd.iso**](https://mega.nz/#F!L1dCkAzY!Z788JUP1Sbknm-WnmhzBHA) [(commit)](https://github.com/reactos/reactos/pull/1171/commits/dfeefe74bf2aa06102f64264c84da7b0d54e2f9c)

It's on MEGA.NZ. If the file is not found, maybe I'm updating it right now.
**To call BSOD, double press LCtrl+Scroll Lock** (works even in text-mode setup)

### Additional screenshots
all
bsod 1
bsod 2

### CODE REVIEW

**Comment by osdeverr** [comment]
Holy crap, this is awesome.

**Comment by katahiromz** [comment]
Can you join our JIRA? https://jira.reactos.org
Can you see ReactOS Coding Style? https://reactos.org/wiki/Coding_Style
Don't use Tab characters. Replace every Tab with a space. Indentation should be 4-letter width.

**Comment by feel-the-dz3n** [comment]
> Can you join our JIRA? https://jira.reactos.org
> Can you see ReactOS Coding Style? https://reactos.org/wiki/Coding_Style
> Don't use Tab characters. Replace every Tab with a space. Indentation should be 4-letter width.

I never seen this, sorry. Going to rework everything.

**Comment by gigaherz** [comment]
First, no need to close a PR, you can update it with more commits, or even force-push a complete new version replacing the old one.

Second, I always found the smiley face to be a bit condescending, maybe we can use a version of the ReactOS logo, like, with the "electrons" fallen down on the ground... if someone wants to make that icon happen.

**Comment by katahiromz** [comment]
Replacing Tabs can be automatically by your text editor.

**Comment by gigaherz** [comment]
Made this, it's quick&dirty, but this is what I was picturing mostly: 



If you want to try it with that image, here's the 16-color bmp file: https://1drv.ms/u/s!AvOILJoyO9_xjmTG4v4yr8eNQn7c

**Comment by feel-the-dz3n** [comment]
> Made this, it's quick&dirty, but this is what I was picturing mostly:
> 
> 
> 
> If you want to try it with that image, here's the 16-color bmp file: https://1drv.ms/u/s!AvOILJoyO9_xjmTG4v4yr8eNQn7c

I will try later.

**Comment by Matriks404** [comment]
As long as it includes the same information as old bluescreen and doesn't introduce any problems in displaying it I am in favour for simplified version here but I would increase padding of space between  ':(' and text below  to make it more readable and maybe add another empty line between description and code error below.

**Comment by blackcrack** [comment]
lil' question, was there not speak about to have a possible for theming SBOD to make it possible for themeing.. and .. by the way, this is not Win10 ... only as hint ..

**Comment by osdeverr** [comment]
@gigaherz, the one you sent looks pretty 2000ish. I'd rather have :-( as the thing.

**Comment by gigaherz** [comment]
Everyone's going to have an opinion. Mine is that I hated Microsoft for having that ":(", the two times I have seen it, so I would prefer not to have that sad smiley in ReactOS. Personally i'd rather keep the existing BSOD rather than have the ":(" there. 🤷‍♂️

**Comment by blackcrack** [comment]
> 
> 
> @gigaherz, the one you sent looks pretty 2000ish. I'd rather have :-( as the thing.

naaa, this is a reactos ball with the rings.. where has fallen apart

and look, this is it, why should it be themeable .. anyone like a other screen..
picture x/y (or topleft, topmiddle, topright, mr,mm,ml, bottomleft,bm,br... , 
Text and background color i guess, this should be enough ..

**Comment by HBelusca** [comment]
Definitively nice workart. Just that the sad smiley looks childish, that's why I don't like Windows 8+ BSODs. It may be replaced by something else.
Also, please think about having a compile-time option to enable or disable this feature  (and reverting to good'ol' BSOD), or maybe some setting value from registry..........

**Comment by gigaherz** [comment]
@blackcrack I don't think themeing belongs in the BSOD screen. It's meant to be as simple as possible, so that there's the least chance the BSOD fails to display. If it was made themeable, then the BSOD itself could crash trying to apply the theme. I'd say even @HBelusca's idea of having a registry key, doesn't belong, but at least since it's JUST one toggle, it can be read at boot and kept in memory without having to worry about keeping a lot of data in memory all the time "in case it crashes".

**Comment by blackcrack** [comment]
(joke) crash in crash, we make it better as MS, right *lol* (/joke)

**Comment by feel-the-dz3n** [comment]
I said that we can create gradient from 16 colors (blue -> white right now) in palette to change colors. So we can implement color changing in registry.

**Comment by HBelusca** [comment]
Some ppl did also this:

**Comment by HBelusca** [comment]
> Wow. Seems like Windows has better bootvid, which can normally draw over images.

Use InbvBitBlt (that's how we draw the boot logo). But the image is still in 16-colour, using an optimized colour palette.

**Comment by feel-the-dz3n** [comment]
accidentally closed ...

File: sdk/include/reactos/mc/bugcodes.mc [hunk 1]
```diff
@@ -1,1711 +1,1703 @@
-MessageIdTypedef=ULONG
```
[the remark this answers is not in the document]
Comment (feel-the-dz3n) [reply sdk/include/reactos/mc/bugcodes.mc]: My editor made line endings like in Windows. Now they again Unix (LF).

**Comment by ThFabba** [comment]
I'm skeptical. When I tried to make the BSoD more useful a while back, it was already short on space. So adding a pointless logo seems counterproductive.

https://github.com/ThFabba/reactos/commit/6a9f172b76bd11f763598c16e5d47299eec1971c
useful-bsod

The QR code is a good idea, but we'd have to figure out what data we want it to contain... and how much data we can reliably fit in there in the first place. And then of course it has to be implemented ;)

**Comment by feel-the-dz3n** [comment]
> 
> [ThFabba@6a9f172](https://github.com/ThFabba/reactos/commit/6a9f172b76bd11f763598c16e5d47299eec1971c)
> useful-bsod
> 

It looks nice and useful, I even like it more than mine.
>
> When I tried to make the BSoD more useful a while back, it was already short on space. 
>
Also what happened to your attempt to implement it?

**Comment by ThFabba** [comment]
> Also what happened to your attempt to implement it?

The usual: I was working on 7000 things at the same time, then had no time to finish 6995 of them.
The current state is that both the backtrace generation code and the mc file changes are kind of hacky, and it needs to be tested with additional types of bug checks.
Feel free to improve upon it or integrate it with your changes... whatever makes sense.

**Comment by gigaherz** [comment]
This is about as much information as I can fit in a QR code with my phone still capable of parsing it. bigger codes don't seem to work.

(qr code contains some random stack trace I took from my reactos/logs/ folder)
qrcode 2

**Comment by HBelusca** [comment]
> The current state is that both the backtrace generation code  [...]

And where's your code @ThFabba ? In one of your github branches?

**Comment by binarymaster** [comment]
@gigaherz we could use a special format for storing backtrace more efficiently, for example having a dictionary of module names, and a list of offsets.

And after that compress it with gzip! :smirk:  QR code will become much smaller.

**Comment by gigaherz** [comment]
@binarymaster Ideally the QR code should be readable text, so that people can scan with their phones and get something meaningful. This QR code is 81x81 squares in size, and is about as big as it can be, since even this size takes a while to parse with my phone, so I believe we would need to have even smaller size, which would make it impossible to have detailed information in a QR code, even gzipping.

**Comment by Mouvedia** [comment]
related: https://jira.reactos.org/browse/CORE-8285

**Comment by gedmurphy** [comment]
I think a mixture of all 3 ideas would be a good way to go. 
- Yaroslav's original idea, colours, etc
- giga's image instead of the sad face
- a QR code encompassing the extra info Thomas put forward in his POC

For the regular development branch, there's probably an argument to not bother with the QR and display the stack trace directly, as Thomas did.

**Comment by blackcrack** [comment]
QR-Code is a cool thing, the users can make Screenshoots or photos and can send directly more information's as in the BSOD direct display's
loveit

best regards
Blacky

**Comment by learn-more** [comment]
@HBelusca  he linked the commit...

**Comment by katahiromz** [comment]
He is under cens@rsh!p, can't go to JIRA.

**Comment by HBelusca** [comment]
> an black/white QR is maybe better ...

I agree.

**Comment by feel-the-dz3n** [comment]
> > an black/white QR is maybe better ...
> 
> I agree.

There's no black color in the palette. Will it work with blue instead of black?

**Comment by HBelusca** [comment]
I would then suggest to reserve in the palette one entry for black and one for white, and all the other ones for the different blue tones.

**Comment by feel-the-dz3n** [comment]
> I would then suggest to reserve in the palette one entry for black and one for white, and all the other ones for the different blue tones.

Good idea! I will try.

**Comment by feel-the-dz3n** [comment]
image

**Comment by learn-more** [comment]
Let's solve this in steps?
First step:
Make the bsod more usable with a stacktrace on it.
We can always spice it up with a QR code later

**Comment by blackcrack** [comment]
it is possible to mirroring in, the created qr-code picture behind save at systemdrive root ?
let us think further herewith it is possible to have a background-color via htmlcode and 2-3 different color for text.. just as idea .. and the BSOD picture, like the Ball could also mirroring in.. and can be the same technick like the QR-code.. therewith could it maybe also possible to give y/x variable of the space to display .. in easy way.. or something .. therewith it is also possible to use different bsod-pictures too and change the colors if the peoples want modding the System in his way .. so could themes possible.
Just a idea ..

File: ntoskrnl/ke/bug.c [hunk 2]
```diff
@@ -28,6 +28,8 @@ ULONG KeBugCheckCount = 1;
 ULONG KiHardwareTrigger;
 PUNICODE_STRING KiBugCheckDriver;
 ULONG_PTR KiBugCheckData[5];
+CHAR* AdditionalInformation;
+ULONG ErrorScreenStyle = 0;
```
Comment (ThFabba) [inline ntoskrnl/ke/bug.c]: These globals are missing the component prefix (most likely `Ki` here)

File: ntoskrnl/ke/bug.c [hunk 3]
```diff
@@ -571,6 +589,19 @@ KeBugCheckUnicodeToAnsi(IN PUNICODE_STRING Unicode,
...
 }
 
+
+VOID
+NTAPI
+AppendAdditionalData(IN CHAR* NewData,
+                     IN BOOLEAN DisplayIt)
+{
+    /* Add new info to AdditionalInformation */
+    sprintf(AdditionalInformation, NewData);
```
Comment (ThFabba) [inline ntoskrnl/ke/bug.c]: This makes no sense. `AdditionalInformation` is never initialized to anything but a null pointer.

File: ntoskrnl/ke/bug.c [hunk 4]
```diff
@@ -634,13 +665,106 @@ KiDumpParameterImages(IN PCHAR Message,
...
 }
 
+VOID
+NTAPI
+SetUpBlueScreen(IN ULONG BackgroundColor)
+{
+    PVOID EmojiResource = NULL;
+    ULONG PositionY = 75; // 109 default
+    
+	if (ErrorScreenStyle >= 1)
```
Comment (ThFabba) [inline ntoskrnl/ke/bug.c]: Broken indent (several times in this function)

File: sdk/include/reactos/mc/bugcodes.mc [hunk 5]
```diff
@@ -79,17 +79,15 @@ Severity=Warning
...
 .
 
 MessageId=0x80
 Severity=Warning
 Facility=System
 SymbolicName=BUGCODE_ID_DRIVER
 Language=English
-The problem seems to be caused by the following file:
-
+What failed:
```
Comment (ThFabba) [inline sdk/include/reactos/mc/bugcodes.mc]: Most definitely not an improvement.

File: sdk/include/reactos/mc/bugcodes.mc [hunk 6]
```diff
@@ -79,17 +79,15 @@ Severity=Warning
 Facility=System
 SymbolicName=BUGCHECK_MESSAGE_INTRO
 Language=English
-A problem has been detected and ReactOS has been shut down to prevent damage
-to your computer.
+Your PC ran into a problem and needs to restart.
```
Comment (ThFabba) [inline sdk/include/reactos/mc/bugcodes.mc]: I'm not sure what's supposed to be better about this string.

File: sdk/include/reactos/mc/bugcodes.mc [hunk 7]
```diff
@@ -109,14 +107,8 @@ Facility=System
...
-If this is a new installation, ask your hardware or software manufacturer
-for any ReactOS updates you might need.
-
-If problems continue, disable or remove any newly installed hardware
-or software. Disable BIOS memory options such as caching or shadowing.
-If you need to use Safe Mode to remove or disable components, restart
-your computer, press F8 to select Advanced Startup Options, and then
-select Safe Mode.
+If this is a new installation, ask your hardware or software
+manufacturer for any ReactOS updates you might need.
```
Comment (ThFabba) [inline sdk/include/reactos/mc/bugcodes.mc]: Why move the line break?

File: sdk/include/reactos/mc/bugcodes.mc [hunk 8]
```diff
@@ -1127,8 +1119,8 @@ memory. Try changing video adapters.
 
 Disable or remove any newly installed hardware and drivers. Disable or
 remove any newly installed software. If you need to use Safe Mode to
-remove or disable components, restart your computer, press F8 to select
-Advanced Startup Options, and then select Safe Mode.
+remove or disable components, restart your computer, press F8 to
+select Advanced Startup Options, and then select Safe Mode.
```
Comment (ThFabba) [inline sdk/include/reactos/mc/bugcodes.mc]: Ditto

File: ntoskrnl/ke/bug.c [hunk 9]
```diff
@@ -650,84 +774,147 @@ KiDisplayBlueScreen(IN ULONG MessageId,
...
+            NextFrame = Frame[0];
+ 
+            if ((ULONG_PTR)FramePc > (ULONG_PTR)MmHighestUserAddress &&
+            KiPcToFileHeader(FramePc, &LdrEntry, FALSE, &InSystem))
+            {
+                KeBugCheckUnicodeToAnsi(&LdrEntry->BaseDllName,
+                                        AnsiName,
+                                        sizeof(AnsiName));
+                FramePc = (PVOID)((ULONG_PTR)FramePc - (ULONG_PTR)LdrEntry->DllBase);
+                sprintf(Buffer, "%p <%s:%p>\r\n", Frame, AnsiName, FramePc);
```
Comment (ThFabba) [inline ntoskrnl/ke/bug.c]: I must have either been really lazy or copied this from somewhere else -- `sprintf` is not allow in new code, please use an ntstrsafe equivalent (e.g. RtlStringCbPrintfA).

File: sdk/include/reactos/mc/bugcodes.mc (hunk 6 above)
Comment (feel-the-dz3n) [reply sdk/include/reactos/mc/bugcodes.mc]: `ReactOS ran into a problem.`

File: sdk/include/reactos/mc/bugcodes.mc (hunk 7 above)
Comment (feel-the-dz3n) [reply sdk/include/reactos/mc/bugcodes.mc]: Original string will look like
```
If this is a new installation, ask your hardware or software manufacturer
 for any ReactOS updates you might need.
```

File: ntoskrnl/ke/bug.c (hunk 9 above)
Comment (feel-the-dz3n) [reply ntoskrnl/ke/bug.c]: No problem

**Comment by njlyf2011** [comment]
How about add multi-language support?

**Comment by binarymaster** [comment]
Not everything in a single PR :sweat_smile:

**Comment by njlyf2011** [comment]
I just made a suggestion, not asking the author must to add this feature...

**Comment by feel-the-dz3n** [comment]
> I just made a suggestion, not asking the author must to add this feature...

There is LANGUAGE in bugcodes.mc, but as I see it's unused
image

**Comment by Cpt-Derp** [comment]
Was asking because " *** STOP: " represents the actual hexadecimal error code. What's "Stop code"?

**Comment by feel-the-dz3n** [comment]
> Was asking because " *** STOP: " represents the actual hexadecimal error code. What's "Stop code"?

it's MessageId, should be Bug code ID

**Comment by Cpt-Derp** [comment]
Ohh that makes sense. Thanks for clarifying.

File: sdk/include/reactos/mc/bugcodes.mc (hunk 6 above)
Comment (blackcrack) [reply sdk/include/reactos/mc/bugcodes.mc]: "Huston, we have a Problem !
we need a  boot "
then with a bit charme ;)

**Comment by feel-the-dz3n** [comment]
Woah @osdeverr 
image

**Comment by njlyf2011** [comment]
What is the font of ":(" ?
I recommend using free fonts to avoid copyright issues.

**Comment by njlyf2011** [comment]
My design (via GIMP and ReactOS Paint)
default

**Comment by blackcrack** [comment]
there should give a theme-possibility with a ini-file.

y/x the Logo picture %win
y/x the qr-code picture
y/x the Textframe (h1,h2/p)
Text Color
Background color

in %theme%

in %systemroot%\Resources\Themes\bsod\

therewith could it also possible for modding for any user.
Like we see it is a diversity need .

and @njlyf2011 we are not MSWin10 *bg* 

we need a possibility to modding the bsod..
therewith could is be really pretty well a exefile where create a normal window with one button[test bsod], where read the bsod.ini and display the bsod screen in the window, nothing extra, easy read the bsod.ini. The Admin can edit the bsod.ini with notepad and can check the configuration in the "dispbsod".exe (Display BSOD) . Therewith have any Admin (or owner) of the System the possible for modding the System-BSOD.
and let show the

**Comment by ThFabba** [comment]
Most of these points have already been mentioned either directly or indirectly, but to be clear, let me summarize the constraints of this work again:
* A bug check is an exceptional situation and the code to handle it must be stable and minimal. The system is already malfunctioning at that point, so there is no time for games if we want to achieve anything useful.
* Debug output cannot be written to disk in case of a crash. Disk I/O is not available during bug check handling. The code to write crash dumps uses a very complicated work-around, which we do not yet have.
* An ini file parser does not belong in the kernel (excessive maintenance burden, unnecessary attack surface). Any configuration must be in the registry and must be read at boot time rather than during bug check handling (the registry is not available at this point).
* Because we have no crash dump support yet and limited live debugging support, our bugchecks, when in doubt, should be _more_ technical and less fancy than Windows's. Otherwise we're working against the point, which is to identify and fix the issue that caused the crash.
* And finally, our target is Windows Server 2003. It is acceptable to implement additional features where it makes sense, but they should not be a distraction from the main goal or we'll never achieve that goal. In particular, large amounts of code that incur additional maintenance burden are a bad idea.

Arguing most of these points seems fruitless (especially if you have no kernel development experience), so please think twice before trying, and if you really think you must, take the discussion to ROS-Dev or IRC.

Here's what I would conclude from that (these are more arguable, but please make a good case so that we avoid blowing up this discussion more than necessary):
* If we decide to add a stack trace or a QR code to the blue screen, it should not be customizable or optional. These things are there to ease debugging, so making them optional is missing the point.
* Offering some customization options is acceptable, but the options should be simple, and compile-time switches may be more appropriate than run-time ones.
* In particular, we should not offer the ability to load a custom image from disk. That falls under the "high-maintenance code" case above and wastes RAM. If someone really wants a different image they can make compile-time changes or use a resource editor (which is how people already successfully make these changes on Windows).

**Comment by osdeverr** [comment]
@njlyf2011, your version has a couple of major flaws. The BSOD is 640*480*4 while yours is some 16:9 resolution, the font is impossible to bring in, the blue is too bright and kinda burns my eyes.

**Comment by njlyf2011** [comment]
Emmmmm
I will modify it ……

**Comment by feel-the-dz3n** [comment]
> Emmmmm
> I will modify it ……

Don't forget that image size and font size are unchageable, and QR-code's size will be large because it will contain a lot of information.

**Comment by feel-the-dz3n** [comment]
>
> Most of these points have already been mentioned either directly or indirectly, but to be clear, let me summarize the constraints of this work again:
> 
At the moment we are trying to achieve stability and flexibility in what we are doing (I'm working with @osdeverr ). We want to slightly change the concept - to achieve as little text as possible, display the basic tech information and display a large QR code in which will be the most necessary for debugging. I also have an assumption that the style can be changed in the command line of the loader. QR-Code is very stable for now. With it's help we can display A HUGE amount of information.

**Comment by njlyf2011** [comment]
2nd version:

default

**Comment by njlyf2011** [comment]
The text and QR are too dense, which may make some people feel uncomfortable.

**Comment by njlyf2011** [comment]
I think the content of QR code should be a URL that describes the current error situation and not the details.

1) Excessive content may cause the low-profile phone to scan the QR code too slowly.
2) If the QR code pattern is too dense, it will cause some people's discomfort.

**Comment by feel-the-dz3n** [comment]
> I think the content of QR code should be a URL that describes the current error situation and not the details.
> 
> 1. Excessive content may cause the low-profile phone to scan the QR code too slowly.
> 2. If the QR code pattern is too dense, it will cause some people's discomfort.

This would be nice for Windows, but not for ReactOS, because we are talking about debug information. But idea is good, I would like to see it in ReactOS 1.0 :rofl:

**Comment by njlyf2011** [comment]
If this is the case, I feel that we may not need a QR code. We just need to move ":(" and "Your PC ... restart" up to display more the information.

**Comment by feel-the-dz3n** [comment]
> If this is the case, I feel that we may not need a QR code. We just need to move ":(" and "Your PC ... restart" up to display more the information.

Enough to change ErrorScreenStyle to display old bsod

**Comment by blackcrack** [comment]
> 
> 
> I think the content of QR code should be a URL that describes the current error situation and not the details.
> 
>     1. Excessive content may cause some phone to scan the QR code too slowly.
> 
>     2. If the QR code pattern is too dense, it will cause some people's discomfort.

read the @ThFabba text : https://github.com/reactos/reactos/pull/1171#issuecomment-449626162

imho mus be the qr-code more bigger as the text, it is unnecessary if the "user"
feel comfortableness , more importable it is to have the Programmer more as possible information for bugfixing, this picture : 
x

and this makes logical sense , the user (who be not technical) become informations for explain what's going on and the Technical is hidden in the code, but, viewable via handy and able to make a photo for send it in to the reactos programming team. more do it not need.. 
half/half , one for the user, one for the Technical 50:50 

me disturbs easy the huge Smily, the picture from Giga is more better imho ..
best regards
Blacky

**Comment by blackcrack** [comment]
Hi Giannis  @yagoulas ,
anyone have today a Handy with QR-Reader, so what .. 
in the qr-code it the explain.. so.. 
you goes aggressive ..

in the forum.. quite nice idea, but the german Mr. Xpert have ban me and have remove the 
posting who can prove to be i am not the one who is guilty , this ban was illegal !
this about the Forum.. and want you exclude the people ? :+1:

**Comment by osdeverr** [comment]
@blackcrack, let me open it up with the fact of your English being terrible and you seemingly writing complete nonsense. Done with the personal stuff, let's move on. The goal is to create an OS, and OSes aren't only used by programmers, are they? Besides, Dz3n is making what I sent right now, albeit edited.

**Comment by blackcrack** [comment]
Hi Giannis @yagoulas ,
anyone have today a Handy with QR-Reader, so what ..
in the qr-code it the explain.. so..
you goes aggressive ..

so, Reactos please delete the Post from Giannis also, he goes agressive to me !

**Comment by ThFabba** [comment]
Enough. Back to topic!

**Comment by blackcrack** [comment]
@ThFabba then delete the Post from Giannis too please !

**Comment by blackcrack** [comment]
Thomas, dann lösch bitte auch das von Giannis,
denn er wird an dieser stelle persönlich..

Thomas, then delete also from Giannis,
because he will be there personally ..

**Comment by blackcrack** [comment]
Thomas, Du solltest mal in den Spiegel schauen und dir überlegen, wo deine correckt heit ist, danke für das mund verbieten..

**Comment by feel-the-dz3n** [comment]
Closed temporary or not

**Comment by WindowxDeveloper** [comment]
As long as old blue screen is kept, this can be used for alternative.

**Comment by feel-the-dz3n** [comment]
The continuation: https://github.com/reactos/reactos/pull/1444

**Pull request closed without merging**